core/src/main/java/org/mule/runtime/core/DefaultMuleEvent.java

```diff
@@ -981,12 +980,7 @@ public Correlation getCorrelation() {
 
   @Override
   public String getCorrelationId() {
-    return getCorrelation().getId().orElse(getExecutionContext().getSourceCorrelationId().orElse(getMessage().getUniqueId()));
-  }
-
-  @Override
-  public boolean hasSourceCorrelation() {
-    return getExecutionContext().getSourceCorrelationId().isPresent();
+    return getCorrelation().getId().orElse(getExecutionContext().getCorrelationId().orElse(getMessage().getUniqueId()));
```

Comment:
If this logic needs to be here then fine, but ideally we'd encapsulate some of this inside Correlation object wouldn't we if possible?